File: app/src/main/java/com/kjjejones42/audiobookplayer/display/DisplayListAdapter.java
package com.kjjejones42.audiobookplayer.display;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.content.Intent;
import android.graphics.Bitmap;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.kjjejones42.audiobookplayer.AudioBook;
import com.kjjejones42.audiobookplayer.R;
import com.kjjejones42.audiobookplayer.database.AudiobookDao;
import com.kjjejones42.audiobookplayer.database.AudiobookDatabase;
import com.kjjejones42.audiobookplayer.player.PlayActivity;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Objects;
import java.util.concurrent.TimeUnit;


public class DisplayListAdapter extends RecyclerView.Adapter<DisplayListAdapter.MyViewHolder> {

    private DisplayListViewModel model;
    private RecyclerView rcv;
    final private DisplayListActivity activity;
    private int selectedPos = RecyclerView.NO_POSITION;
    private final View.OnClickListener onClickListener = v -> {
        int position = rcv.getChildLayoutPosition(v);
        List<ListItem> items = model.getListItems().getValue();
        assert items != null;
        if (items.get(position).getHeadingOrItem() == ListItem.TYPE_ITEM) {
            AudioBook book = ((ListItem.AudioBookContainer) items.get(position)).book;
            selectedPos = position;
            notifyItemChanged(position);
            startAudioBook(book);
        }
    };
    private List<ListItem> currentItems;
    private final View.OnLongClickListener onLongClickListener = v -> {
        List<ListItem> items = model.getListItems().getValue();
        String[] statuses = AudioBook.getStatusMap().values().toArray(new String[0]);
        assert items != null;
        AudiobookDao dao = AudiobookDatabase.getInstance(v.getContext()).audiobookDao();
        ListItem.AudioBookContainer container = ((ListItem.AudioBookContainer) items.get(rcv.getChildLayoutPosition(v)));
        String bookId = container.book.displayName;
        new AlertDialog.Builder(v.getContext())
                .setSingleChoiceItems(statuses, dao.getStatus(bookId), (dialog, which) -> {
                    AudioBook book = dao.findByName(bookId);
                    book.setStatus(which);
                    dao.update(book);
                    dialog.dismiss();
                }).setTitle("Choose this book's status.")
                .setNegativeButton("Cancel", (dialog, which) -> dialog.dismiss())
                .show();
        return false;
    };

    DisplayListAdapter(@NonNull DisplayListViewModel model, @NonNull RecyclerView rcv, DisplayListActivity activity) {
        this.model = model;
        this.activity = activity;
        this.rcv = rcv;
        model.getListItems().observe(activity, this::selectivelyNotify);
        setHasStableIds(true);
    }

    private void startAudioBook(AudioBook book) {
//        AudiobookDatabase.getInstance(activity).audiobookDao()
//                .getAllAndObserve().removeObservers(activity);
        Intent intent = new Intent(activity, PlayActivity.class);
        intent.putExtra(DisplayListActivity.INTENT_PLAY_FILE, book.displayName);
        intent.putExtra(DisplayListActivity.INTENT_START_PLAYBACK, true);
        activity.startActivity(intent);
    }


    @SuppressLint("NotifyDataSetChanged")
    private void selectivelyNotify(List<ListItem> newItems) {
        List<ListItem> oldItems = currentItems;
        currentItems = newItems;
        if (newItems.equals(oldItems)) {
            return;
        }
        if (oldItems == null) {
            notifyItemRangeInserted(0, newItems.size());
            return;
        }
        if (oldItems.size() == newItems.size()) {
            notifyDataSetChanged();
            return;
        }
        boolean remove = oldItems.size() > newItems.size();
        List<ListItem> larger = remove ? oldItems : newItems;
        List<ListItem> smaller = remove ? newItems : oldItems;
        List<Integer> indexes = new ArrayList<>();
        for (int i = 0; i < larger.size(); i++) {
            if (!smaller.contains(larger.get(i))) {
                indexes.add(i);
            }
        }
        indexes.sort((o1, o2) -> o2 - o1);
        if (remove) {
            for (int i : indexes) {
                notifyItemRemoved(i);
            }
        } else {
            for (int i : indexes) {
                notifyItemInserted(i);
            }
        }
        rcv.scrollToPosition(0);
    }

    private List<ListItem> getItems() {
        if (currentItems != null) {
            return currentItems;
        }
        return Objects.requireNonNull(model.getListItems().getValue());
    }

    @Override
    public long getItemId(int position) {
        return getItems().get(position).getId();
    }

    @Override
    public int getItemViewType(int position) {
        return getItems().get(position).getHeadingOrItem();
    }


    void filter(String filterTerm) {
        List<AudioBook> books = model.getSavedBooks().getValue();
        if (books != null && filterTerm != null) {
            List<AudioBook> filtered = new ArrayList<>();
            for (AudioBook book : books) {
                if (book.toString().toUpperCase().contains(filterTerm.toUpperCase())) {
                    filtered.add(book);
                }
            }
            model.setFilteredListItems(filtered);
        }
    }

    @NonNull
    @Override
    public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v;
        if (viewType == ListItem.TYPE_HEADING) {
            v = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.display_list_group, parent, false);
            return new MyViewHolder(v, false);
        } else {
            v = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.display_list_item, parent, false);
            v.setOnClickListener(onClickListener);
            v.setOnLongClickListener(onLongClickListener);
            return new MyViewHolder(v, true);
        }
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onBindViewHolder(@NonNull MyViewHolder holder, int position) {
        List<ListItem> items = getItems();
        assert items != null;
        switch (items.get(position).getHeadingOrItem()) {
            case ListItem.TYPE_ITEM:
                AudioBook book = ((ListItem.AudioBookContainer) items.get(position)).book;
                holder.duration.setText(msToReadableDuration(book.getTotalDuration()));
                holder.textView.setText(book.displayName);
                holder.artist.setText(book.author);
                holder.textView.setSelected(selectedPos == position);
                new Thread(() -> {
                    Bitmap thumbnail = book.getThumbnail(activity);
                    holder.v.post(() -> {
                        holder.image.setImageBitmap(thumbnail);
                        holder.image.setVisibility(View.VISIBLE);
                    });
                }).start();
                break;

            case ListItem.TYPE_HEADING:
                String title = ((ListItem.Heading) items.get(position)).getHeadingTitle();
                holder.textView.setText(title);
                break;
        }
    }

    private String msToReadableDuration(long ms) {
        long minutes = TimeUnit.MILLISECONDS.toMinutes(ms) % 60;
        long hours = TimeUnit.MILLISECONDS.toHours(ms);
        if (hours > 0) {
            return String.format(Locale.getDefault(), "%dh %02dm", hours, minutes);
        }
        return String.format(Locale.getDefault(), "%02dm", minutes);
    }

    @Override
    public int getItemCount() {
        return getItems().size();
    }

    static class MyViewHolder extends RecyclerView.ViewHolder {
        final TextView textView;
        final View v;
        TextView artist;

        TextView duration;
        ImageView image;

        MyViewHolder(View v, boolean isItem) {
            super(v);
            this.v = v;
            if (isItem) {
                textView = v.findViewById(R.id.listItemText);
                image = v.findViewById(R.id.listImageView);
                artist = v.findViewById(R.id.artist);
                duration = v.findViewById(R.id.listItemDuration);
                image.setVisibility(View.INVISIBLE);
            } else {
                textView = (TextView) v;
            }
        }
    }
}

File: app/src/main/java/com/kjjejones42/audiobookplayer/display/DisplayListAdapter.kt
package com.kjjejones42.audiobookplayer.display

import android.annotation.SuppressLint
import android.app.AlertDialog
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnLongClickListener
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.kjjejones42.audiobookplayer.AudioBook
import com.kjjejones42.audiobookplayer.AudioBookStatus
import com.kjjejones42.audiobookplayer.R
import com.kjjejones42.audiobookplayer.database.AudiobookDatabase.Companion.getInstance
import com.kjjejones42.audiobookplayer.display.DisplayListAdapter.MyViewHolder
import com.kjjejones42.audiobookplayer.display.ListItem.AudioBookContainer
import com.kjjejones42.audiobookplayer.display.ListItem.Heading
import com.kjjejones42.audiobookplayer.player.PlayActivity
import java.util.Locale
import java.util.concurrent.TimeUnit

private fun msToReadableDuration(ms: Long): String {
    val minutes = TimeUnit.MILLISECONDS.toMinutes(ms) % 60
    val hours = TimeUnit.MILLISECONDS.toHours(ms)
    if (hours > 0) {
        return String.format(Locale.getDefault(), "%dh %02dm", hours, minutes)
    }
    return String.format(Locale.getDefault(), "%02dm", minutes)
}

class DisplayListAdapter internal constructor(
    private var model: DisplayListViewModel,
    private var rcv: RecyclerView,
    private val activity: DisplayListActivity
) :
    RecyclerView.Adapter<MyViewHolder>() {
    private var selectedPos = RecyclerView.NO_POSITION
    private val onClickListener = View.OnClickListener {
        val position = rcv.getChildLayoutPosition(it)
        model.listItems.value?.takeIf { it[position].type == ListItemType.ITEM }?.let {
            selectedPos = position
            notifyItemChanged(position)
            startAudioBook((it[position] as AudioBookContainer).book)
        }
    }
    private var currentItems: List<ListItem>? = null
    private val onLongClickListener = OnLongClickListener { v ->
        model.listItems.value?.let {
            val statuses = AudioBookStatus.entries.map { it.displayName }.toTypedArray()
            val dao = getInstance(v.context).audiobookDao()
            val container = (it[rcv.getChildLayoutPosition(v)] as AudioBookContainer)
            val bookId = container.book.displayName
            AlertDialog.Builder(v.context)
                .setSingleChoiceItems(statuses, dao.getStatus(bookId)) { dialog, which ->
                    dao.findByName(bookId)?.let {
                        it.setStatus(AudioBookStatus.entries[which])
                        dao.update(it)
                    }
                    dialog.dismiss()
                }.setTitle("Choose this book's status.")
                .setNegativeButton("Cancel") { dialog, _ -> dialog.dismiss() }
                .show()
        }
        false
    }

    init {
        model.listItems.observe(activity) { this.selectivelyNotify(it) }
        setHasStableIds(true)
    }

    private fun startAudioBook(book: AudioBook) {
//        AudiobookDatabase.getInstance(activity).audiobookDao()
//                .getAllAndObserve().removeObservers(activity);
        val intent = Intent(activity, PlayActivity::class.java)
        intent.putExtra(DisplayListActivity.INTENT_PLAY_FILE, book.displayName)
        intent.putExtra(DisplayListActivity.INTENT_START_PLAYBACK, true)
        activity.startActivity(intent)
    }


    @SuppressLint("NotifyDataSetChanged")
    private fun selectivelyNotify(newItems: List<ListItem>) {
        val oldItems = currentItems
        currentItems = newItems
        if (newItems == oldItems) {
            return
        }
        if (oldItems == null) {
            notifyItemRangeInserted(0, newItems.size)
            return
        }
        if (oldItems.size == newItems.size) {
            notifyDataSetChanged()
            return
        }
        val remove = oldItems.size > newItems.size
        val larger = if (remove) oldItems else newItems
        val smaller = if (remove) newItems else oldItems
        larger.indices
            .filter { !smaller.contains(larger[it]) }
            .sortedDescending()
            .forEach {
                if (remove) {
                    notifyItemRemoved(it)
                } else {
                    notifyItemInserted(it)
                }
            }
        rcv.scrollToPosition(0)
    }

    private val items: List<ListItem>
        get() {
            currentItems?.let { return it }
            return model.listItems.value ?: emptyList()
        }

    override fun getItemId(position: Int): Long {
        return items[position].id
    }

    override fun getItemViewType(position: Int): Int {
        return items[position].type.value
    }


    fun filter(filterTerm: String?) {
        model.savedBooks.value?.takeIf { filterTerm != null }?.let {
            val filtered = it
                .filter { it.toString().uppercase().contains(filterTerm!!.uppercase()) }
                .toList()
            model.setFilteredListItems(filtered)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
        if (viewType == ListItemType.HEADING.value) {
            val v = LayoutInflater.from(parent.context)
                .inflate(R.layout.display_list_group, parent, false)
            return MyViewHolder(v, false)
        } else {
            val v = LayoutInflater.from(parent.context)
                .inflate(R.layout.display_list_item, parent, false)
            v.setOnClickListener(onClickListener)
            v.setOnLongClickListener(onLongClickListener)
            return MyViewHolder(v, true)
        }
    }

    @SuppressLint("SetTextI18n")
    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
        when (items[position].type) {
            ListItemType.ITEM -> {
                val book = (items[position] as AudioBookContainer).book
                holder.duration!!.text = msToReadableDuration(book.totalDuration)
                holder.textView!!.text = book.displayName
                holder.artist!!.text = book.author
                holder.textView!!.isSelected = selectedPos == position
                Thread {
                    val thumbnail = book.getThumbnail(activity)
                    holder.v.post {
                        holder.image!!.setImageBitmap(thumbnail)
                        holder.image!!.visibility = View.VISIBLE
                    }
                }.start()
            }

            ListItemType.HEADING -> {
                val title = (items[position] as Heading).headingTitle
                holder.textView!!.text = title
            }
        }
    }

    override fun getItemCount(): Int {
        return items.size
    }

    class MyViewHolder(val v: View, isItem: Boolean) : RecyclerView.ViewHolder(v) {
        var textView: TextView? = null
        var artist: TextView? = null

        var duration: TextView? = null
        var image: ImageView? = null

        init {
            if (isItem) {
                textView = v.findViewById(R.id.listItemText)
                image = v.findViewById(R.id.listImageView)
                artist = v.findViewById(R.id.artist)
                duration = v.findViewById(R.id.listItemDuration)
                image!!.visibility = View.INVISIBLE
            } else {
                textView = v as TextView
            }
        }
    }
}
